include/triton/Conversion/TritonGPUToLLVM/Utility.h

```diff
@@ -29,89 +30,249 @@
+                                        std::forward<Args>(args)...);
+  }
+  template <typename... Args> LLVM::ICmpOp icmp_ugt(Args &&...args) {
+    return builder.create<LLVM::ICmpOp>(loc, LLVM::ICmpPredicate::ugt,
+                                        std::forward<Args>(args)...);
+  }
+  template <typename... Args> LLVM::ICmpOp icmp_uge(Args &&...args) {
+    return builder.create<LLVM::ICmpOp>(loc, LLVM::ICmpPredicate::uge,
+                                        std::forward<Args>(args)...);
+  }
+  template <typename... Args> LLVM::SelectOp select(Args &&...args) {
+    return builder.create<LLVM::SelectOp>(loc, std::forward<Args>(args)...);
+  }
+  template <typename... Args> LLVM::AddressOfOp address_of(Args &&...args) {
+    return builder.create<LLVM::AddressOfOp>(loc, std::forward<Args>(args)...);
+  }
+  mlir::gpu::BarrierOp barrier() {
+    return builder.create<mlir::gpu::BarrierOp>(loc);
+  }
+  template <typename... Args> LLVM::UndefOp undef(Args &&...args) {
+    return builder.create<LLVM::UndefOp>(loc, std::forward<Args>(args)...);
+  }
+  template <typename... Args> LLVM::ZeroOp null(Args &&...args) {
+    return builder.create<LLVM::ZeroOp>(loc, std::forward<Args>(args)...);
+  }
+  template <typename... Args> LLVM::CallOp call(Args &&...args) {
+    return builder.create<LLVM::CallOp>(loc, std::forward<Args>(args)...);
+  }
+  // Constants
+  Value int_val(short bitwidth, int64_t val) {
+    Type ty = builder.getIntegerType(bitwidth);
+    return builder.create<LLVM::ConstantOp>(loc, ty,
+                                            builder.getIntegerAttr(ty, val));
+  }
+  Value i1_val(int64_t val) { return int_val(1, val); }
+  Value true_val() { return int_val(1, true); }
+  Value false_val() { return int_val(1, false); }
+  Value f16_val(float v) {
+    auto type = type::f16Ty(builder.getContext());
+    return builder.create<LLVM::ConstantOp>(loc, type,
+                                            builder.getF16FloatAttr(v));
+  }
+  Value f32_val(float v) {
+    auto type = type::f32Ty(builder.getContext());
+    return builder.create<LLVM::ConstantOp>(loc, type,
+                                            builder.getF32FloatAttr(v));
+  }
+  Value f64_val(double v) {
+    auto type = type::f64Ty(builder.getContext());
+    return builder.create<LLVM::ConstantOp>(loc, type,
+                                            builder.getF64FloatAttr(v));
+  }
+  Value i8_val(int64_t val) { return int_val(8, val); }
+  Value i16_val(int64_t val) { return int_val(16, val); }
+  Value i32_val(int64_t val) { return int_val(32, val); }
+  Value i64_val(int64_t val) { return int_val(64, val); }
+  Value tid_val() {
+    Value tid =
+        builder.create<::mlir::gpu::ThreadIdOp>(loc, ::mlir::gpu::Dimension::x);
+    Type i32_ty = builder.getIntegerType(32);
+    return builder.create<arith::IndexCastOp>(loc, i32_ty, tid);
+  }
+
+  Location loc;
+  RewriterBase &builder;
```

Comment:
Reference data members is an anti-pattern because it disables copy-assignment. Also I don't think we need a rewriter here, just any `OpBuilder` should do.
```suggestion
  OpBuilder *builder;
```